File: ruby/grade-school/school.rb
class School

  def initialize
    @roster = Roster.new
  end

  def add(name, n)
    @roster.add name, n
    nil
  end

  def grade(n)
    @roster.find_grade n
  end

  def to_hash
    @roster.to_hash
  end

end

class Roster

  def initialize
    @db = Hash.new
  end

  def add(name, n)
    find_or_create_grade(n).push(name).sort!
    nil
  end

  def find_grade(n)
    find_or_create_grade(n).freeze
  end

  def to_hash
    Hash[@db.sort].each_value{ |grade| grade.freeze }
  end

  private

  def find_or_create_grade(n)
    @db[n] ||= []
  end

end

File: elixir/grade-school/school.exs
defmodule School do
  @moduledoc """
  Simulate students in a school.

  Each student is in a grade.
  """

  @doc """
  Add a student to a particular grade in school.
  """
  @spec add(Dict.t, String.t, pos_integer) :: Dict.t
  def add(db, name, grade) do
    db
    |> Dict.update(grade,
                   [name],
                   fn (students_in_grade) ->
                     [name | students_in_grade]
                   end)
  end

  @doc """
  Return the names of the students in a particular grade.
  """
  @spec grade(Dict.t, pos_integer) :: [String]
  def grade(db, grade) do
    db
    |> Dict.fetch(grade)
    |> case do
        {:ok, names} ->
          names
        :error ->
          []
      end
  end

  @doc """
  Sorts the school by grade and name.
  """
  @spec sort(Dict) :: Dict.t
  def sort(db), do: db # already sorted...?
end
